### PR TITLE
Hash module

### PR DESCRIPTION
**Pull request by drknzz**
## Introduced changes

- Create `starknet_py.hash` module
- **starknet.js** uses similiar [structure](https://github.com/0xs34n/starknet.js/blob/develop/src/utils/hash.ts)

##

- [x] This PR contains breaking changes

| Function | Old Path | New Path |
| ------------- | ------------- | -- |
| compute_address | starknet_py.net.models.address  | starknet_py.hash.address |
| get_selector_from_name |  starknet_py.cairo.selector  | starknet_py.hash.selector |
| get_storage_var_address |  starknet_py.cairo.storage  | starknet_py.hash.storage |
| transaction hashes |  starknet_py.utils.crypto.transaction_hash  | starknet_py.hash.transaction |
| _starknet_keccak |  starknet_py.cairo.utils  | starknet_py.hash.utils |
| |   | |
| compute_hash_on_elements |  starknet_py.utils.crypto.facade  | starknet_py.hash.utils |
| message_signature |  starknet_py.utils.crypto.facade  | starknet_py.hash.utils |
| pederse_hash |  starknet_py.utils.crypto.facade  | starknet_py.hash.utils |

### CODE REVIEW

File: starknet_py/hash/utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from starknet_py.common import int_from_bytes
+from starknet_py.constants import MASK_250
+from starknet_py.utils.crypto.facade import use_cpp_variant
+
+
+def _starknet_keccak(data: bytes) -> int:
+    """
+    A variant of eth-keccak that computes a value that fits in a StarkNet field element.
+    """
+    return int_from_bytes(keccak(data)) & MASK_250
```
Comment (cptartur) [inline starknet_py/hash/utils.py:18]: Do we want to keep this a private function?

File: starknet_py/hash/utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def pedersen_hash(left: int, right: int) -> int:
```
Comment (cptartur) [inline starknet_py/hash/utils.py:21]: This could use a docstring

File: starknet_py/hash/utils.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    if use_cpp_variant():
+        return cpp_hash(left, right)
+    return default_hash(left, right)
```
Comment (cptartur) [inline starknet_py/hash/utils.py:24]: Let's add some kind of test for it.

File: starknet_py/hash/utils.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        h(h(h(h(0, data[0]), data[1]), ...), data[n-1]), n).
+
+    The hash is initialized with 0 and ends with the data length appended.
+    The length is appended in order to avoid collisions of the following kind:
+    H([x,y,z]) = h(h(x,y),z) = H([w, z]) where w = h(x,y).
+    """
+    return functools.reduce(pedersen_hash, [*data, len(data)], 0)
+
+
+def message_signature(msg_hash, priv_key, seed: Optional[int] = 32) -> ECSignature:
```
Comment (cptartur) [inline starknet_py/hash/utils.py:39]: Docstring

**Review by cptartur — COMMENTED** [review]
I like the changes. 👍 

I'm not 100% sold on the `utils` name. Technically I think it'd make sense to name this file `hash` but then we'd have `from ...hash.hash` imports which I don't really like. 

If there are some alternatives to `utils` you've considered I'd be interesting in seeing them though.

File: starknet_py/hash/utils.py (hunk 3 above)
Comment (drknzz) [reply starknet_py/hash/utils.py:24]: Brother 💀 I was just moving stuff around in this pr.

File: starknet_py/hash/utils.py (hunk 1 above)
Comment (drknzz) [reply starknet_py/hash/utils.py:18]: Any reasons for why we should make it public?

Comment (cptartur) [reply starknet_py/hash/utils.py:18]: I wonder, probably no.

**Comment by drknzz** [comment]
> I like the changes. 👍
> 
> I'm not 100% sold on the `utils` name. Technically I think it'd make sense to name this file `hash` but then we'd have `from ...hash.hash` imports which I don't really like.
> 
> If there are some alternatives to `utils` you've considered I'd be interesting in seeing them though.

How about `common`? They are not that common though. And I think utils is better than that

**Comment by cptartur** [comment]
> > I like the changes. 👍
> > I'm not 100% sold on the `utils` name. Technically I think it'd make sense to name this file `hash` but then we'd have `from ...hash.hash` imports which I don't really like.
> > If there are some alternatives to `utils` you've considered I'd be interesting in seeing them though.
> 
> How about `common`? They are not that common though. And I think utils is better than that

I also prefer `utils` to `common`.

File: starknet_py/hash/contract_class.py [hunk 5]
```diff
@@ -2,10 +2,9 @@
 from typing import List
```
Comment (war-in) [inline starknet_py/hash/contract_class.py:2]: Why this file is named `contract_class`?

File: starknet_py/hash/address.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+from typing import Sequence
```
Comment (war-in) [inline starknet_py/hash/address.py:1]: Do you think having two `address` files is ok?

**Review by war-in — COMMENTED** [review]
I like the structure 🔥 
Although `utils` is not the best name, it is located in the `hash` directory (fair enough).

File: starknet_py/hash/address.py (hunk 6 above)
Comment (drknzz) [reply starknet_py/hash/address.py:1]: As they are in different places and contain other logic, I'd say yes

File: starknet_py/hash/contract_class.py (hunk 5 above)
Comment (drknzz) [reply starknet_py/hash/contract_class.py:2]: It cannot be called `class.py` as it's some internal name 😿

Comment (war-in) [reply starknet_py/hash/contract_class.py:2]: I would name it `class_hash`. It is a very well-known name in StarkNet and import 
```from starknet.py.hash.class_hash import compute_class_hash```
isn't too bad

**Review by war-in — APPROVED** [review]
Just rename `contract_class` to `class_hash` and it's perfect 🥇

File: starknet_py/hash/contract_class.py (hunk 5 above)
Comment (drknzz) [reply starknet_py/hash/contract_class.py:2]: @cptartur, you mentioned the repeating names in module path above, what do you think?

**Review by cptartur — APPROVED** [review]

File: starknet_py/hash/contract_class.py (hunk 5 above)
Comment (cptartur) [reply starknet_py/hash/contract_class.py:2]: I think it's okay in this case. It'd be worse if it was `hash.hash`.

Comment (cptartur) [reply starknet_py/hash/contract_class.py:2]: Let's name it as @war-in suggested.

Comment (drknzz) [reply starknet_py/hash/contract_class.py:2]: Okay, I will do it in next pr though

**Pull request merged**